internal/statesync/statesync.go

```diff
@@ -0,0 +1,314 @@
+
+type StateSyncer struct {
+	// statesync configuration
+	dbConfig *DBConfig
+
+	// directory to store snapshots and chunks
+	// same as the snapshot store directory
+	// as we allow to reuse the received snapshots
+	snapshotsDir string
+
+	// trusted snapshot providers for verification - cometbft rfc servers
+	snapshotProviders []*rpchttp.HTTP
+
+	// State syncer state
+	inProgress bool
+	snapshot   *Snapshot
+	chunks     map[uint32]bool // Chunks received till now
+	rcvdChunks uint32          // Number of chunks received till now
+
+	// Logger
+	log log.Logger
+}
+
+func NewStateSyncer(ctx context.Context, cfg *DBConfig, snapshotDir string, providers []string, logger log.Logger) *StateSyncer {
+
+	ss := &StateSyncer{
+		dbConfig:     cfg,
+		snapshotsDir: snapshotDir,
+		log:          logger,
+	}
+
+	for _, s := range providers {
+		clt, err := ChainRPCClient(s)
+		if err != nil {
+			logger.Error("Failed to create rpc client", zap.String("server", s), zap.Error(err))
+			return nil
+		}
+		ss.snapshotProviders = append(ss.snapshotProviders, clt)
+	}
+
+	// Ensure that the snasphot directory exists and is empty
+	if err := os.RemoveAll(snapshotDir); err != nil {
+		logger.Error("Failed to delete snapshot directory", zap.String("dir", snapshotDir), zap.Error(err))
+		return nil
+	}
+	if err := os.MkdirAll(snapshotDir, 0755); err != nil {
+		logger.Error("Failed to create snapshot directory", zap.String("dir", snapshotDir), zap.Error(err))
+		return nil
+	}
+
+	return ss
+}
+
+// OfferSnapshot checks if the snapshot is valid and kicks off the state sync process
+// accepted snapshot is stored on disk for later processing
+func (ss *StateSyncer) OfferSnapshot(snapshot *Snapshot) error {
+	ss.log.Info("Offering snapshot", zap.Int64("height", int64(snapshot.Height)), zap.Uint32("format", snapshot.Format), zap.String("App Hash", fmt.Sprintf("%x", snapshot.SnapshotHash)))
+
+	// Check if we are already in the middle of a snapshot
+	if ss.inProgress {
+		return ErrStateSyncInProgress
+	}
+
+	// Validate the snapshot
+	err := ss.validateSnapshot(*snapshot)
+	if err != nil {
+		return err
+	}
+
+	ss.snapshot = snapshot
+	ss.inProgress = true
+	ss.chunks = make(map[uint32]bool, snapshot.ChunkCount)
+	ss.rcvdChunks = 0
+	return nil
+}
+
+// ApplySnapshotChunk accepts a chunk and stores it on disk for later processing if valid
+// If all chunks are received, it starts the process of restoring the database
+func (ss *StateSyncer) ApplySnapshotChunk(ctx context.Context, chunk []byte, index uint32) (bool, bool, error) {
+	if !ss.inProgress {
+		return false, false, ErrStateSyncNotInProgress
+	}
+
+	// Check if the chunk has already been applied
+	applied, ok := ss.chunks[index]
+	if ok && applied {
+		return false, false, nil
+	}
+
+	// Check if the chunk index is valid
+	if index >= ss.snapshot.ChunkCount {
+		ss.log.Error("Invalid chunk index", zap.Uint32("index", index), zap.Uint32("chunkCount", ss.snapshot.ChunkCount))
```

Comment:
can you use hyphenated log info? `chunk-count` instead of `chunkCount`. Just to keep it consistent.